File: crates/yutha-passport/src/passport.rs
//! [`Passport`] and [`PassportBuilder`].
//!
//! Mirrors `Passport` from
//! [`/spec/passport/passport-v1.proto`](../../../spec/passport/passport-v1.proto).
//!
//! A passport is content-addressable: hash of its canonical serialization
//! with `agent_signature` cleared is its identifier. Signatures and
//! seal-style metadata are excluded from canonical bytes (same pattern as
//! receipts).

use crate::declarations::{CapabilityDeclaration, ResourceDeclaration};
use crate::error::{PassportError, Result};
use crate::tier::PassportTier;
use yutha_core::{AgentId, PublicKey, Signature, SpecVersion, SwarmId, Timestamp};
use yutha_crypto::canonical::Canonical;
use yutha_crypto::Result as CryptoResult;
use yutha_proto::Message;

/// A signed identity manifest. What an agent presents to a registry.
#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub struct Passport {
    /// Spec version this passport was authored under.
    pub spec_version: SpecVersion,
    /// Stable agent identifier. UUID v7.
    pub agent_id: AgentId,
    /// Swarm this passport is valid for (passports are single-swarm).
    pub swarm_id: SwarmId,
    /// Public key the agent uses to sign envelopes and receipts.
    pub agent_public_key: PublicKey,
    /// Human-readable owner identifier (org / team / individual).
    pub owner: String,
    /// Framework the agent is built on (e.g. `"langgraph"`, `"autogen"`).
    pub framework: String,
    /// Framework version string.
    pub framework_version: String,
    /// What the agent claims it can do. NOT authority — see capability spec.
    pub capabilities: Vec<CapabilityDeclaration>,
    /// Constitution version this agent commits to obey.
    pub accepted_constitution_version: String,
    /// Tier mirroring conformance tiers.
    pub tier: PassportTier,
    /// Declared resource budget.
    pub resources: ResourceDeclaration,
    /// When issued.
    pub issued_at: Timestamp,
    /// When this passport stops being valid. Optional in closed swarms;
    /// required in open/hybrid (sybil mitigation — re-registration cost).
    pub expires_at: Option<Timestamp>,
    /// Default model provider for A2 attribution.
    pub default_model_provider: String,
    /// Default model name.
    pub default_model_name: String,
    /// Self-signature over canonical bytes (signature cleared).
    pub agent_signature: Option<Signature>,
}

impl Passport {
    /// Builder for constructing a passport.
    pub fn builder() -> PassportBuilder {
        PassportBuilder::default()
    }

    /// Check whether this passport has expired against the given monotonic
    /// timestamp. Returns false for passports without `expires_at`.
    pub fn is_expired_at(&self, now: &Timestamp) -> bool {
        self.expires_at
            .as_ref()
            .map(|e| e.monotonic_ns <= now.monotonic_ns)
            .unwrap_or(false)
    }
}

impl Canonical for Passport {
    /// Produce canonical bytes for content-addressing and signature
    /// verification.
    ///
    /// Goes through [`Passport::to_canonical_proto`] (which clears
    /// `agent_signature` and `extensions`) and then encodes via
    /// `prost::Message::encode_to_vec()`. With prost's tag-sorted field
    /// encoding plus `btree_map(["."])` in `yutha-proto`'s build (so the
    /// `bounds` map on each [`CapabilityDeclaration`] encodes with sorted
    /// keys), the result is bytewise deterministic both within a single Rust
    /// process and across spec-conforming implementations.
    fn canonical_bytes(&self) -> CryptoResult<Vec<u8>> {
        Ok(self.to_canonical_proto().encode_to_vec())
    }
}

// -----------------------------------------------------------------------------
// Builder
// -----------------------------------------------------------------------------

/// Builder for [`Passport`].
///
/// The agent's signing key produces `agent_signature` at the very end via
/// [`PassportBuilder::sign`]; callers MAY construct unsigned passports for
/// pre-flight validation but the registry will reject them.
#[derive(Debug, Default)]
pub struct PassportBuilder {
    spec_version: Option<SpecVersion>,
    agent_id: Option<AgentId>,
    swarm_id: Option<SwarmId>,
    agent_public_key: Option<PublicKey>,
    owner: String,
    framework: String,
    framework_version: String,
    capabilities: Vec<CapabilityDeclaration>,
    accepted_constitution_version: String,
    tier: PassportTier,
    resources: ResourceDeclaration,
    issued_at: Option<Timestamp>,
    expires_at: Option<Timestamp>,
    default_model_provider: String,
    default_model_name: String,
}

impl PassportBuilder {
    /// Required: spec version.
    pub fn spec_version(mut self, v: SpecVersion) -> Self {
        self.spec_version = Some(v);
        self
    }
    /// Required: agent id.
    pub fn agent_id(mut self, id: AgentId) -> Self {
        self.agent_id = Some(id);
        self
    }
    /// Required: swarm id.
    pub fn swarm_id(mut self, id: SwarmId) -> Self {
        self.swarm_id = Some(id);
        self
    }
    /// Required: agent's public key.
    pub fn agent_public_key(mut self, pk: PublicKey) -> Self {
        self.agent_public_key = Some(pk);
        self
    }
    /// Optional: human-readable owner.
    pub fn owner(mut self, owner: impl Into<String>) -> Self {
        self.owner = owner.into();
        self
    }
    /// Optional: framework + version.
    pub fn framework(mut self, name: impl Into<String>, version: impl Into<String>) -> Self {
        self.framework = name.into();
        self.framework_version = version.into();
        self
    }
    /// Optional: add a capability declaration.
    pub fn declares(mut self, decl: CapabilityDeclaration) -> Self {
        self.capabilities.push(decl);
        self
    }
    /// Required: constitution version the agent commits to.
    pub fn accepted_constitution_version(mut self, v: impl Into<String>) -> Self {
        self.accepted_constitution_version = v.into();
        self
    }
    /// Optional: passport tier (default: Minimal).
    pub fn tier(mut self, tier: PassportTier) -> Self {
        self.tier = tier;
        self
    }
    /// Optional: resource declaration.
    pub fn resources(mut self, r: ResourceDeclaration) -> Self {
        self.resources = r;
        self
    }
    /// Required: issued-at timestamp.
    pub fn issued_at(mut self, t: Timestamp) -> Self {
        self.issued_at = Some(t);
        self
    }
    /// Optional: expiration. Required in open/hybrid swarms by topology.
    pub fn expires_at(mut self, t: Timestamp) -> Self {
        self.expires_at = Some(t);
        self
    }
    /// Optional: default model provider for A2 attribution.
    pub fn default_model(mut self, provider: impl Into<String>, name: impl Into<String>) -> Self {
        self.default_model_provider = provider.into();
        self.default_model_name = name.into();
        self
    }

    /// Build an unsigned passport. Useful for pre-flight validation.
    pub fn build_unsigned(self) -> Result<Passport> {
        Ok(Passport {
            spec_version: self
                .spec_version
                .ok_or(PassportError::MissingField("spec_version"))?,
            agent_id: self
                .agent_id
                .ok_or(PassportError::MissingField("agent_id"))?,
            swarm_id: self
                .swarm_id
                .ok_or(PassportError::MissingField("swarm_id"))?,
            agent_public_key: self
                .agent_public_key
                .ok_or(PassportError::MissingField("agent_public_key"))?,
            owner: self.owner,
            framework: self.framework,
            framework_version: self.framework_version,
            capabilities: self.capabilities,
            accepted_constitution_version: self.accepted_constitution_version,
            tier: self.tier,
            resources: self.resources,
            issued_at: self
                .issued_at
                .ok_or(PassportError::MissingField("issued_at"))?,
            expires_at: self.expires_at,
            default_model_provider: self.default_model_provider,
            default_model_name: self.default_model_name,
            agent_signature: None,
        })
    }

    /// Build and sign with the supplied signing key. The signing key's
    /// public counterpart MUST match the `agent_public_key` field — this is
    /// re-checked here.
    pub fn sign(self, signing_key: &yutha_crypto::SigningKey) -> Result<Passport> {
        let mut p = self.build_unsigned()?;
        if p.agent_public_key != signing_key.public() {
            return Err(PassportError::SelfSignatureInvalid);
        }
        let bytes = p.canonical_bytes().map_err(PassportError::Crypto)?;
        let sig = signing_key.sign_message(&bytes);
        p.agent_signature = Some(sig);
        Ok(p)
    }
}

// `Default for PassportTier` is derived in tier.rs with `#[default]` on `Minimal`.

// -----------------------------------------------------------------------------
// Verification
// -----------------------------------------------------------------------------

impl Passport {
    /// Verify the passport's self-signature against the inlined public key.
    /// Returns Ok on success; [`PassportError::SelfSignatureInvalid`] on
    /// mismatch or if the signature is missing.
    pub fn verify_self_signature(&self) -> Result<()> {
        let sig = self
            .agent_signature
            .as_ref()
            .ok_or(PassportError::SelfSignatureInvalid)?;
        let bytes = self.canonical_bytes().map_err(PassportError::Crypto)?;
        yutha_crypto::sign::verify(&self.agent_public_key, &bytes, sig)
            .map_err(|_| PassportError::SelfSignatureInvalid)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use yutha_crypto::sign::generate_keypair;

    fn fixture() -> (yutha_crypto::SigningKey, AgentId) {
        (generate_keypair(), AgentId::new())
    }

    fn build_signed(key: &yutha_crypto::SigningKey, agent_id: AgentId) -> Passport {
        Passport::builder()
            .spec_version(SpecVersion::parse("1.0.0").unwrap())
            .agent_id(agent_id)
            .swarm_id(SwarmId::new())
            .agent_public_key(key.public())
            .owner("test")
            .framework("test-framework", "0.1.0")
            .accepted_constitution_version("1.0.0")
            .tier(PassportTier::Minimal)
            .issued_at(Timestamp::now())
            .sign(key)
            .unwrap()
    }

    #[test]
    fn builder_requires_required_fields() {
        let result = Passport::builder().build_unsigned();
        assert!(result.is_err());
    }

    #[test]
    fn signed_passport_verifies() {
        let (key, agent_id) = fixture();
        let p = build_signed(&key, agent_id);
        assert!(p.verify_self_signature().is_ok());
    }

    #[test]
    fn passport_with_mismatched_public_key_rejects_at_sign() {
        let (key1, agent_id) = fixture();
        let key2 = generate_keypair();
        let result = Passport::builder()
            .spec_version(SpecVersion::parse("1.0.0").unwrap())
            .agent_id(agent_id)
            .swarm_id(SwarmId::new())
            .agent_public_key(key2.public()) // mismatch with signing key
            .accepted_constitution_version("1.0.0")
            .issued_at(Timestamp::now())
            .sign(&key1);
        assert!(matches!(result, Err(PassportError::SelfSignatureInvalid)));
    }

    #[test]
    fn tampered_passport_fails_verification() {
        let (key, agent_id) = fixture();
        let mut p = build_signed(&key, agent_id);
        p.owner = "tampered".into();
        assert!(matches!(
            p.verify_self_signature(),
            Err(PassportError::SelfSignatureInvalid)
        ));
    }

    #[test]
    fn unsigned_passport_fails_verification() {
        let (key, agent_id) = fixture();
        let mut p = build_signed(&key, agent_id);
        p.agent_signature = None;
        assert!(matches!(
            p.verify_self_signature(),
            Err(PassportError::SelfSignatureInvalid)
        ));
    }

    #[test]
    fn is_expired_when_past_monotonic() {
        let (key, agent_id) = fixture();
        let mut p = build_signed(&key, agent_id);
        // Force an expires_at strictly less than the issued_at monotonic.
        p.expires_at = Some(
            Timestamp::new(
                "2020-01-01T00:00:00Z".into(),
                p.issued_at.monotonic_ns.saturating_sub(1),
            )
            .unwrap(),
        );
        let now = Timestamp::now();
        assert!(p.is_expired_at(&now));
    }

    #[test]
    fn passport_without_expires_at_never_expires() {
        let (key, agent_id) = fixture();
        let p = build_signed(&key, agent_id);
        let far_future = Timestamp::new("3030-01-01T00:00:00Z".into(), u64::MAX).unwrap();
        assert!(!p.is_expired_at(&far_future));
    }
}

File: sdks/python/src/yutha/models/passport.py
"""Passport ergonomic model + supporting types.

Mirrors ``/spec/passport/passport-v1.proto`` and the Rust ``Passport``
in ``crates/yutha-passport/src/passport.rs``.
"""

from __future__ import annotations

from enum import IntEnum

from pydantic import BaseModel, ConfigDict, Field

from yutha._proto import common_pb2
from yutha._proto.passport import passport_v1_pb2 as proto
from yutha.canonical import canonical_bytes as _canonical_bytes
from yutha.crypto import Signer, VerificationFailed, verify
from yutha.identity import AgentId, PublicKey, Signature, SwarmId, Timestamp


class PassportTier(IntEnum):
    """Conformance tier. Wire numbers match the proto enum directly."""

    MINIMAL = 1
    STANDARD = 2
    VERIFIABLE = 3


class CapabilityDeclaration(BaseModel):
    """Declared capability (NOT authority — see capability spec).

    ``bounds`` is a sorted dict on the wire (the Rust side uses BTreeMap
    so canonical bytes are deterministic). On the Python side we let the
    user pass a plain dict; we sort it on the way to proto.
    """

    model_config = ConfigDict(frozen=True)

    kind: str
    resource_tags: list[str] = Field(default_factory=list)
    bounds: dict[str, str] = Field(default_factory=dict)
    description: str = ""

    @classmethod
    def from_proto(cls, p: proto.CapabilityDeclaration) -> CapabilityDeclaration:
        return cls(
            kind=p.kind,
            resource_tags=list(p.resource_tags),
            bounds=dict(p.bounds),
            description=p.description,
        )

    def to_proto(self) -> proto.CapabilityDeclaration:
        out = proto.CapabilityDeclaration(
            kind=self.kind,
            resource_tags=list(self.resource_tags),
            description=self.description,
        )
        # Protobuf maps don't preserve insertion order; deterministic
        # encoding sorts them at write time. We just copy.
        for k, v in self.bounds.items():
            out.bounds[k] = v
        return out


class ResourceDeclaration(BaseModel):
    """Declared resource budget. Pure scalars — round-trips cleanly."""

    model_config = ConfigDict(frozen=True)

    max_concurrent_actions: int = 0
    max_messages_per_minute: int = 0
    max_tool_calls_per_hour: int = 0
    max_usd_per_day_cents: str = ""
    max_memory_bytes: int = 0

    @classmethod
    def from_proto(cls, p: proto.ResourceDeclaration) -> ResourceDeclaration:
        return cls(
            max_concurrent_actions=p.max_concurrent_actions,
            max_messages_per_minute=p.max_messages_per_minute,
            max_tool_calls_per_hour=p.max_tool_calls_per_hour,
            max_usd_per_day_cents=p.max_usd_per_day_cents,
            max_memory_bytes=p.max_memory_bytes,
        )

    def to_proto(self) -> proto.ResourceDeclaration:
        return proto.ResourceDeclaration(
            max_concurrent_actions=self.max_concurrent_actions,
            max_messages_per_minute=self.max_messages_per_minute,
            max_tool_calls_per_hour=self.max_tool_calls_per_hour,
            max_usd_per_day_cents=self.max_usd_per_day_cents,
            max_memory_bytes=self.max_memory_bytes,
        )


class Passport(BaseModel):
    """A signed identity manifest — what an agent presents at swarm
    join.

    Construct, then ``await passport.sign(signer)`` to attach
    ``agent_signature`` over canonical bytes. The returned Passport is
    a new immutable instance; the original is unchanged.
    """

    model_config = ConfigDict(frozen=True)

    spec_version: str
    agent_id: AgentId
    swarm_id: SwarmId
    agent_public_key: PublicKey
    owner: str = ""
    framework: str = ""
    framework_version: str = ""
    capabilities: list[CapabilityDeclaration] = Field(default_factory=list)
    accepted_constitution_version: str
    tier: PassportTier = PassportTier.MINIMAL
    resources: ResourceDeclaration = Field(default_factory=ResourceDeclaration)
    issued_at: Timestamp
    expires_at: Timestamp | None = None
    default_model_provider: str = ""
    default_model_name: str = ""
    agent_signature: Signature | None = None

    # -------------------------------------------------------------------------
    # Round-trip
    # -------------------------------------------------------------------------

    @classmethod
    def from_proto(cls, p: proto.Passport) -> Passport:
        return cls(
            spec_version=p.spec_version.value,
            agent_id=AgentId.from_proto(p.agent_id),
            swarm_id=SwarmId.from_proto(p.swarm_id),
            agent_public_key=PublicKey.from_proto(p.agent_public_key),
            owner=p.owner,
            framework=p.framework,
            framework_version=p.framework_version,
            capabilities=[CapabilityDeclaration.from_proto(c) for c in p.capabilities],
            accepted_constitution_version=p.accepted_constitution_version,
            tier=PassportTier(p.tier),
            resources=ResourceDeclaration.from_proto(p.resources),
            issued_at=Timestamp.from_proto(p.issued_at),
            expires_at=Timestamp.from_proto(p.expires_at) if p.HasField("expires_at") else None,
            default_model_provider=p.default_model_provider,
            default_model_name=p.default_model_name,
            agent_signature=Signature.from_proto(p.agent_signature)
            if p.HasField("agent_signature")
            else None,
        )

    def to_proto(self) -> proto.Passport:
        out = proto.Passport(
            spec_version=common_pb2.Version(value=self.spec_version),
            agent_id=self.agent_id.to_proto(),
            swarm_id=self.swarm_id.to_proto(),
            agent_public_key=self.agent_public_key.to_proto(),
            owner=self.owner,
            framework=self.framework,
            framework_version=self.framework_version,
            capabilities=[c.to_proto() for c in self.capabilities],
            accepted_constitution_version=self.accepted_constitution_version,
            # proto.Passport's `tier` parameter is typed as
            # `proto.PassportTier | str | None`; we pass an int from our
            # IntEnum. Same wire numbers, nominally distinct types.
            tier=self.tier.value,  # type: ignore[arg-type]
            resources=self.resources.to_proto(),
            issued_at=self.issued_at.to_proto(),
            default_model_provider=self.default_model_provider,
            default_model_name=self.default_model_name,
        )
        if self.expires_at is not None:
            out.expires_at.CopyFrom(self.expires_at.to_proto())
        if self.agent_signature is not None:
            out.agent_signature.CopyFrom(self.agent_signature.to_proto())
        return out

    # -------------------------------------------------------------------------
    # Canonical bytes + signing
    # -------------------------------------------------------------------------

    def canonical_bytes(self) -> bytes:
        """Canonical wire encoding with ``agent_signature`` and
        ``extensions`` cleared — what the agent signature is over and
        what the content-address hashes."""
        p = self.to_proto()
        p.ClearField("agent_signature")
        p.ClearField("extensions")
        return _canonical_bytes(p)

    async def sign(self, signer: Signer) -> Passport:
        """Return a copy of this Passport with ``agent_signature`` set.

        The signer's public key MUST match ``agent_public_key``; this
        is verified before signing. ``signer.sign_message`` is awaited
        — for the default :class:`yutha.crypto.InProcessSigner` this
        completes immediately, for cloud-KMS-backed signers it's one
        network round-trip.
        """
        if signer.public_key().value != self.agent_public_key.value:
            raise ValueError("signer does not match agent_public_key")
        sig = await signer.sign_message(self.canonical_bytes())
        return self.model_copy(update={"agent_signature": sig})

    def verify_self_signature(self) -> None:
        """Verify ``agent_signature`` against ``agent_public_key`` and
        the canonical bytes.

        Raises ``VerificationFailed`` if the signature is missing,
        malformed, or doesn't match. Returns ``None`` on success.
        """
        if self.agent_signature is None:
            raise VerificationFailed("passport has no agent_signature to verify")
        verify(self.agent_public_key, self.canonical_bytes(), self.agent_signature)
